lib/chai/utils/overwriteMethod.js

```diff
@@ -52,12 +52,24 @@ module.exports = function (ctx, name, method) {
   if (_method && 'function' === typeof _method)
     _super = _method;
 
-  var fn = function () {
-    var keep_ssfi = flag(this, 'keep_ssfi');
-    var old_ssfi = flag(this, 'ssfi');
-    if (!keep_ssfi && old_ssfi)
-      flag(this, 'ssfi', fn);
+  var overwritingMethodWrapper = function () {
+    // If proxy protection is disabled and this overwriting assertion hasn't
+    // been overwritten again by yet another assertion, then use this method
+    // wrapper as the starting point for removing implementation frames from the
+    // stack trace of a failed assertion.
+    //
+    // Note: If this assertion has been overwritten, and thus the `keep_ssfi`
+    // flag is set, then the overwriting method wrapper is used as the starting
+    // point instead. This prevents the overwriting method wrapper from showing
+    // up in the stack trace since it's invoked before this method wrapper.
+    if (!flag(this, 'keep_ssfi')) {
+      flag(this, 'ssfi', overwritingMethodWrapper);
+    }
 
+    // The `keep_ssfi` flag is set so that if this assertion ends up calling
+    // the overwritten assertion, then the overwritten assertion doesn't attempt
+    // to use itself as the starting point for removing implementation frames
+    // from the stack trace of a failed assertion.
     flag(this, 'keep_ssfi', true);
```

Comment:
Let me see if I've got this right, I'm not sure if I fully understand what this does.

So, we use the SSFI flag to store the function which should indicate where the real stack trace will start in order to remove internal implementation details, right?

Whenever an assertion gets overwritten we must turn the `keep_ssfi` flag to true before calling the old assertion (`_super`) in order to avoid that assertion being used as the start of the stack trace. Right?